# Ulcer Performance Index portfolio optimization with Riskfolio-Lib and Portfolio Optimizer

> This notebook is an illustration of Ulcer Performance Index portfolio optimization in Python, using two libraries:
> - [Riskfolio-Lib](https://riskfolio-lib.readthedocs.io/en/latest/), a native Python library
> - [Portfolio Optimizer](https://portfoliooptimizer.io/), a Web API

> Let's first validate that both libraries output the same optimal portfolio using a toy example.

In [3]:
import numpy as np
import pandas as pd

# Define the (total return) prices for each asset (in EUR):
# - ICE US Treasury Short Bond
# - ICE US Treasury 20+ Year Bond
# - Gold (spot)
# - S&P 500
asset_prices = np.array([[10000,9761.842988170647,9991.809068774734,9840.02312838631,9532.762959697731,9322.352996588435,9471.236332452569,9472.663990702988,9452.237154723576,9637.579878863502,9646.899731128764,9317.581140505525,9378.653317690507,9572.00084584733,9423.42258182162,9389.07181187518,9228.95336954773,9374.322962259243,9376.474650446451,9275.976763499308,9337.019006155242,9060.845325568293,8921.748662927806,8783.689153989166,8829.4099745503,8799.222999478205,8648.717761193602,8317.878709740478,8470.290491354071,8489.796658220841,8364.25012710281,8466.365180557485,8984.883177173533,9287.517646498314,10441.04863320049,10490.170894634288,9607.868765961188,10431.222263003549,10572.006105489168,10053.574339583187,10086.286896963224,9500.008056504585,9477.412893916648,9479.084935521192,9394.711716536582,9160.405991379268,9065.95198309753,8934.897278707656,9315.73827460324,9614.696529909754,9895.942148534443,9962.933097661147,10087.712806677833,10917.787446283935,10954.115149471669,10320.658087587311,10606.9963264759,9856.570325443223,9710.490177924166,10352.541122359764,10072.907391092087,9832.88614809055,9733.935403120235,9480.492644646969,9067.590888127925,9368.797982587113,9326.285147826187,9451.188986787532,9332.95932936371,9986.730606983669,9633.217897435747,10053.085999259085,10425.794698236545,10238.870389793985,10034.874307725784,10101.123835450373,10211.231773257372,10880.983589722353,10719.73382509836,10989.404932518197,10705.634996311901,10443.347854907623,10393.430472078102,10401.120915598167,10239.40019208718,9972.07066695771,10292.200662835723,10554.758901169835,10341.217495749162,10394.28899115896,10336.084533946869,10186.450722918315,10218.05659095122,10015.688480829715,9915.415217029062,9937.170176414746,9809.082504669976,10009.954528524193,9795.55989967335,9815.051442564934,9772.73891564407,9947.553871108707,9911.064077656882,10118.717985950836,10266.571533121256,10761.370305635202,10811.005801145537,10848.642830895522,11152.418291555676,11980.912442602077,12049.94378866197,12589.779157258157,12081.809571385818,12352.53011979406,12112.634626059726,12358.060059193558,12085.291005553956,12105.097479407115,12307.634298587993,12813.56508791692,12452.50585423867,12424.794140078135,12464.585510017272,11930.162225254693,11917.01303171578,12183.520427422738,12253.517918766995,12244.816991391008,12226.18283556755,12201.399545157079,12445.417606001545,12809.69120324448,12930.700271678772,12682.446821288842,12877.630186678247,12763.93517746457,12492.300492394988,12174.505023815911,11980.02115133482,11669.70434188193,11585.224225571017,11612.527238059336,11789.070970306186,11586.49504090704,11458.28443307341,11043.07519588302,11271.31499727645,11189.770592250523,11428.32135268645,11819.495809224523,11880.321333525735,11820.341540964748,11929.173985256271,12022.76435232378,12319.576638223674,12299.874728077151,12229.104109084117,12216.975198814187,12316.994069550285,12545.049373090184,12589.100698893504,12695.731384575876,12473.599499686523,12749.444070899672,12915.245626239106,13109.893519523188,12829.114745647894,13045.00612234689,12771.738668979404,13002.212669728004,13121.3577720766,13210.667848670939,13307.26744301232,12995.599414125645,12925.704773085654,12220.18456573702,12126.58205075232,12369.516548027355],
                        [12528.11144697723,12041.958738500429,12436.663532770406,11642.565619509887,10924.370287925725,10652.387770509453,10885.360710185887,11063.844692312614,11348.269204718572,11733.764444702038,11805.841011428185,11621.836322449646,11322.535252702593,11389.546813334902,11563.495249123021,11265.366234629859,11124.764160261488,11003.577180354701,10846.317013699334,11026.565126067542,11236.013464058806,10892.171914565737,10863.329071226117,11190.810756289091,11170.446275949495,11291.578671174851,11035.99401619919,10789.665851365537,10710.863111410386,10446.759597148493,10540.872822168183,10576.85254170828,11547.758420892285,12093.833769919447,13310.651340124794,15242.174785378827,15805.379246945846,14916.759019882411,14886.15276040921,14737.092424325552,13759.300136086295,12487.204445419007,12571.213174912977,12609.573554814187,12792.435549630545,12771.781025669892,12325.254165332482,12270.701942242758,11995.67900648252,12700.1868142295,13080.649799057204,12846.588637471246,13438.613051023503,15295.9805563484,16185.942760153659,15150.135218498435,16827.935282496233,15279.195058610656,14352.09727044764,15107.23714939391,14148.911778731203,13386.338683158534,13446.74910537199,13101.085649310517,12772.72105617953,13681.094594550834,13307.186879127097,14105.458673144578,15304.057184180812,18405.227605826716,16961.00336035836,18237.876073043037,19565.283529970795,19137.487792515498,18331.546851463554,17684.37742219858,18696.92574957237,21603.453888237065,20978.432354460565,22292.1814066256,21435.510163079438,20415.915919954467,20299.140434552824,20589.899665112756,19838.794315863663,18562.443509990953,19399.460529853153,19846.022317203588,20307.391579012656,19015.463491842173,18306.510062131685,17642.375706943058,17573.167525079105,17241.36986266052,17310.19895619048,16890.72504685616,16338.22951173761,17662.496939464945,17435.22000847701,17605.03644658717,17908.898456653602,18778.69806841778,18669.19900891103,19192.7761702712,20320.746831030163,20873.257761937017,21576.070300338164,22281.73195583488,23644.564476089123,27815.598309449302,26288.740398584865,27804.622807964828,25788.999343614632,25872.967780968007,24321.048134970908,25725.8414703916,25172.877842281985,25573.107550753626,25870.94295099113,26707.92998910918,25945.703941380612,27224.88992146392,28164.899440122375,26919.263858834085,26725.341975853866,27569.628809881808,29537.00939187098,30156.822834424565,29818.922105547525,29217.804712515008,28521.791197711424,27095.663803192256,27185.35002979774,26748.69214811464,27588.051360485293,27183.91832785901,27007.963340606508,26852.85834503228,26530.629279338493,25635.47876584539,26340.809566327327,25790.457358138774,26150.424846515816,25887.744442300325,26043.000093186354,24240.318850330452,23959.58512239014,24497.797640767,24495.40971088594,25846.145773282227,25996.552400067398,25416.506317283987,26015.323070485876,25384.82530206677,25135.51566462356,25517.522466624174,26738.027299601494,26814.275985959717,26651.16205649316,28536.447680199777,28046.578377686143,30124.09686475377,29898.400724090057,30566.155810687767,34249.33810164931,33791.52247925982,32685.68443712577,33037.28676234558,31365.10473399964,34135.0926498745,36732.8712834553,39392.62147756818,40147.32940241727,38412.15858741786,38271.85273984556,37767.89187200634,35764.48017409051,36662.756187592095],
                        [26501.22162451681,28602.933068192273,28533.90964136004,29303.141212734474,31304.902091578835,30925.892176793084,29409.431430025284,30191.99283828399,29567.842934987326,28846.58757287254,28980.797903039627,29857.197304907484,29416.198318785657,30602.96576996069,30639.61381069543,30281.33708853677,30325.657862641598,29857.420031325055,29354.3738307346,29588.691199057903,29882.04679514291,31934.73521542203,33303.85982085266,32347.694204951695,34629.70093292768,37838.02467996976,39035.81278871047,35978.878761285814,34157.59216865411,34807.71200590731,35962.7886603957,35837.03458475786,34452.04212473462,37686.93398686704,34909.23862033754,39001.84507746405,37878.30952276487,43723.9453862981,45885.193718801886,41970.07887996042,40547.98010099495,42168.667306101284,40293.43821651425,40476.012679066145,40800.54485049009,41441.967045778016,42824.44391293949,47695.61230302326,46703.10674387147,47061.7659955467,49771.187255854595,50434.998197786525,53974.02706376135,59793.64754643555,61781.796494072274,54683.57125972378,59885.116272577536,58361.53178785961,59229.47396957929,64866.839622705724,64319.80947820593,59064.17480396795,62158.33185137042,61727.47635934174,62972.526769018215,65094.27685550029,63480.86932988161,69596.65162779728,76483.8641026426,73114.72573449754,74953.8939380436,79300.61586799925,74158.69187412411,80664.60338185275,80241.15587347167,75858.69232481773,76155.03753058903,76552.75379422192,77376.13255357883,80469.47176411937,79663.52769705896,83707.53341683264,80628.12242671172,81000.08995391906,76859.39304671875,74873.79339058878,73735.338644371,76065.06463528833,68485.6946144431,65342.36921284798,55537.79329382154,60345.677716114194,64223.28750865064,59859.45716597916,59150.971855409865,56102.36189755423,53094.32758185888,56406.51067106308,58524.72084460643,57094.8983148101,56696.30522986486,56006.885120945684,58675.7639530529,58544.22579435345,59415.21908492838,58918.00392297798,56653.050037732864,57742.3020972239,60197.07321713233,67936.98163941868,65822.21364910401,67235.54348818482,64134.96167032932,66186.7530916982,63780.17838316235,61036.94510724451,61676.06989690641,60599.768235278316,63191.10429497133,61172.90599927892,59461.878500866915,62047.47884858503,69120.00911940075,66215.0594051325,68710.59956276062,66225.90768659522,72500.25009445798,73593.8138304099,71675.29568747038,72212.54994279225,70819.32382459992,67509.48113384175,67013.02163570313,68722.48923118699,72208.526206699,70960.84892345972,70613.54172570421,68768.70643144609,66338.76020118552,65871.61066927465,67603.63243465732,66233.51970424516,66511.50393051752,65844.00386276576,65881.6997968579,65802.90589006714,65754.9078672233,65480.642266279676,66255.20485710958,67998.34472653664,65367.57716604649,63401.25968203124,62896.13125235936,62503.418156882406,65419.741558203954,65323.104091956244,68215.6565676083,70196.84353199645,70420.69874760466,70266.92108652246,69661.739508407,70804.43656388609,75455.14176400144,78018.50443192125,84400.58280703219,83127.80089023834,82554.2737555161,81028.19074011697,82620.07638001579,87355.29628241716,89376.19386795077,89497.44412081149,95411.73611947392,94604.22189693319,96224.67863256238,101068.42908403196,99904.32082981277,98216.91279844959],
                        [77826.76450231716,77771.33081607298,79578.18253903686,79044.5113586008,77338.9331492533,73181.05278965965,74173.51645263718,74315.73746995458,75586.60876926364,78704.08881089176,81038.1035823385,79426.42698416796,80723.98010536887,83311.11814625066,80092.46763933495,80337.76812212892,82126.72253443325,85952.68128343778,84199.34931809189,80386.28467554077,81603.0530119575,81825.04341575308,81584.59863293036,76510.74302762919,76186.44018665339,70899.08625760608,67252.82290376868,64230.81055442642,68537.97003335693,69569.04177790355,62669.51903438076,62751.659747848586,67443.99143541513,63289.87975552122,59042.125814393155,54934.9311091487,50772.24697880536,50486.82135138543,45724.744803910326,47248.865943618526,51899.71050526833,51603.3326919428,51574.084312809624,55459.1190948808,56921.879098884885,57550.06144176653,55881.84385223278,58354.50199019944,62029.219727645555,61681.7929760662,65448.17230098147,69866.29591753367,71843.3574829008,71521.06148658054,67975.87141495089,68511.79781964108,67214.59043006603,68020.20072711844,69543.56632626946,74148.85181210816,76949.442400336,76874.6224701487,78692.1512815022,76656.57657425861,75458.41421992896,77067.7317061039,75426.52566602397,74892.96551125478,69893.21735636295,69537.19415569205,74393.37797545927,77454.27428180604,81143.21835041196,83254.59213598924,85129.59256661215,88503.80802849642,88893.64662615112,89014.20583081743,91305.2077425054,94879.36589011192,94500.76010122153,94551.19560049151,92355.30896861128,92940.98572370442,92309.52632231684,94539.91811047641,98896.05514937446,105201.13223464359,105037.84824396997,108040.18322546584,105986.37666619815,109743.5654250437,106887.31772892871,108035.15368826993,111874.62150975711,115537.97609960058,116916.94396255445,115170.9850789108,117849.9413763946,119055.78593623087,119398.88550701021,124384.26223806324,126479.52585964437,127336.35448150524,134348.23385479525,138833.64874850327,142894.6467783719,147219.62285193434,150985.32903941572,157282.8117530653,167284.2638313612,171999.01224257285,166588.7223988152,172499.19263576754,165848.83016753444,172751.24902657463,158738.82963575135,154976.8865237053,170887.08168508677,178491.3887256983,170706.24028381234,161744.97407591384,162001.49247448676,165439.49259505555,165818.89649066888,172564.73102956495,173822.42352323892,180052.5033670997,179997.7065377974,179563.5639188777,179750.868118708,191859.1065051954,197396.00564448707,197137.63183494203,208021.6208053113,206432.76803609586,203992.63703500092,201498.63599788529,199362.8543069641,197997.09409226594,196957.44762001888,201343.74990252647,209016.37018431537,211590.62101788144,211374.33669266495,215152.2405480207,211345.366197615,204185.47639462512,209075.46465827982,221065.0807913198,223207.86298911352,229975.795625422,239201.78492016374,242121.92944709002,230714.95582910645,234566.75516919576,211691.8393498436,227899.38281756768,236700.39142713146,245187.21205385542,255501.2803701653,240702.2814460773,252481.1632384848,261369.64998849778,259909.6421637799,268347.1223385362,267645.70067243546,281705.0115287695,283697.7505152901,288256.4966102714,266334.9632688278,233886.44698862042,265810.2209567702,271968.44439677964,275841.6478975753,275408.64066905325,292930.34756277787,287383.8322809532]])

# Define the risk free rate
rf = 0

# Compute the asset arithmetic returns, required for Riskfolio-Lib
df = pd.DataFrame(data=np.transpose(asset_prices))
df.columns = ["ICE US Treasury Short Bond", "ICE US Treasury 20+ Year Bond", "Gold", "S&P 500"]
returns = df.pct_change().dropna()
print(returns)

     ICE US Treasury Short Bond  ICE US Treasury 20+ Year Bond      Gold  \
1                     -0.023816                      -0.038805  0.079306   
2                      0.023558                       0.032777 -0.002413   
3                     -0.015191                      -0.063851  0.026959   
4                     -0.031226                      -0.061687  0.068312   
5                     -0.022072                      -0.024897 -0.012107   
..                          ...                            ...       ...   
173                   -0.023421                      -0.043220 -0.008463   
174                   -0.005378                      -0.003653  0.017129   
175                   -0.054583                      -0.013168  0.050338   
176                   -0.007660                      -0.053045 -0.011518   
177                    0.020033                       0.025116 -0.016890   

      S&P 500  
1   -0.000712  
2    0.023233  
3   -0.006706  
4   -0.021577  
5   -0.

In [4]:
import riskfolio as rp

# Calculate the optimal portfolio with Riskfolio-Lib

## Build the portfolio object
port = rp.Portfolio(returns=returns)

## Estimate the input parameters
method_mu='hist' # Estimate expected returns based on historical data
method_cov='hist' # Estimate covariance matrix based on historical data, not needed
port.assets_stats(method_mu=method_mu, method_cov=method_cov, d=0.94)

## Define the parameters for the portfolio optimization
model='Classic'
rm = 'UCI' # Ulcer Index as risk measure
obj = 'Sharpe' # Maximum Ulcer Performance Index as objective function
hist = True # Use historical scenarios
l = 0 # Risk aversion factor, not used

## Run the optimization
w = port.optimization(model=model, rm=rm, obj=obj, rf=rf, l=l, hist=hist)
print(w)

                                    weights
ICE US Treasury Short Bond     1.150682e-11
ICE US Treasury 20+ Year Bond  2.473364e-01
Gold                           3.264099e-01
S&P 500                        4.262537e-01


In [5]:
import requests

# Calculate the optimal portfolio with Portfolio Optimizer
po_request_body = { 'assets': list(map(lambda x: {'assetPrices': x.tolist()}, asset_prices)), 
                    'riskFreeRate': rf }
response = requests.post('https://api.portfoliooptimizer.io/v1/portfolio/optimization/maximum-ulcer-performance-index', json=po_request_body)
print(response.json())

{'assetsWeights': [9.38765921326449e-10, 0.24733625866959888, 0.32641004564771176, 0.42625369474392305]}


> At this stage, both libraries seem to agree pretty well!

> Let's continue, adding bound constraints (= constraints on maximum portfolio positions).

In [17]:
# Calculate the optimal portfolio with Riskfolio-Lib, adding the following bound constraints:
# ICE US Treasury Short Bond <= 0.1
# ICE US Treasury 20+ Year Bond <= 0.2
# Gold <= 0.5
# S&P 500 <= 0.4

## Build the constraints matrix A and right hand side b such that A*w <= b
A = -np.eye(4)
b = np.array([[-.1], [-0.2], [-0.5], [-0.4]])

## Build the portfolio object
port_constr = rp.Portfolio(returns=returns, ainequality=A, binequality=b)

## Estimate the input parameters
method_mu='hist' # Estimate expected returns based on historical data
method_cov='hist' # Estimate covariance matrix based on historical data, not needed
port_constr.assets_stats(method_mu=method_mu, method_cov=method_cov, d=0.94)

## Define the parameters for the portfolio optimization
model='Classic'
rm = 'UCI' # Ulcer Index as risk measure
obj = 'Sharpe' # Maximum Ulcer Performance Index as objective function
hist = True # Use historical scenarios
l = 0 # Risk aversion factor, not used
rf = 0 # Risk free rate

## Run the optimization
w = port_constr.optimization(model=model, rm=rm, obj=obj, rf=rf, l=l, hist=hist)
print(w)


                                weights
ICE US Treasury Short Bond     0.049233
ICE US Treasury 20+ Year Bond  0.200000
Gold                           0.350767
S&P 500                        0.400000


In [18]:
# Calculate the optimal portfolio with Portfolio Optimizer, adding the same bound constraints
po_request_body = { 'assets': list(map(lambda x: {'assetPrices': x.tolist()}, asset_prices)), 
                    'riskFreeRate': rf, 
                    'constraints': { 'maximumAssetsWeights': list((-b).flat) } }
response = requests.post('https://api.portfoliooptimizer.io/v1/portfolio/optimization/maximum-ulcer-performance-index', json=po_request_body)
print(response.json())

{'assetsWeights': [0.04923287888148928, 0.1999999995903749, 0.35076712159766793, 0.3999999999306333]}


> All good!

> Let's finally check the performances with more assets and more asset prices.

In [6]:
# Dowload the large scale data sets from Riskfolio-Lib
r = requests.get('https://github.com/dcajasn/Riskfolio-Lib/raw/master/examples/assets_data.csv', allow_redirects=True)
open('assets_data.csv', 'wb').write(r.content)

19071905

In [7]:
# Extract 20 assets / 500 price points, corresponding to the maximum API limits for the free usage of Portfolio Optimizer
asset_prices_ls = pd.read_csv("assets_data.csv", index_col='Dates')
asset_prices_tls = asset_prices_ls.iloc[0:500,0:20]
asset_prices_tls_np = np.array(asset_prices_tls).T
# print(asset_prices_tls_list)

# Compute the asset arithmetic returns, required for Riskfolio-Lib
returns_tls = asset_prices_tls.pct_change().dropna()
print(returns_tls)

                   A      AAPL       ABC      ABMD       ABT      ADBE  \
Dates                                                                    
04-01-2000 -0.083842 -0.084321 -0.068261 -0.023945 -0.019714 -0.083890   
05-01-2000 -0.047666  0.014642  0.077584  0.010498 -0.007307  0.019773   
06-01-2000 -0.057057 -0.086530  0.076008  0.001722  0.016572  0.008163   
07-01-2000  0.104035  0.047364  0.130097 -0.005158  0.039857  0.048583   
10-01-2000  0.060575 -0.017588  0.026318  0.142825 -0.022653  0.038610   
...              ...       ...       ...       ...       ...       ...   
26-11-2001  0.045821  0.077129 -0.019895  0.057320  0.001855  0.053695   
27-11-2001  0.028570 -0.017296 -0.004940 -0.026324 -0.004998 -0.019754   
28-11-2001 -0.001852 -0.022400  0.011172 -0.035176  0.005767 -0.063668   
29-11-2001  0.008162 -0.005319  0.019827  0.036458  0.008506  0.048659   
30-11-2001  0.003682  0.043055  0.023221 -0.046734  0.008246 -0.045806   

                 ADM      ADSK       

In [45]:
%%time

# Calculate the optimal "truncated large scale" portfolio with Riskfolio-Lib

## Build the portfolio object
port_tls = rp.Portfolio(returns=returns_tls)

## Estimate the input parameters
method_mu='hist' # Estimate expected returns based on historical data
method_cov='hist' # Estimate covariance matrix based on historical data, not needed
port_tls.assets_stats(method_mu=method_mu, method_cov=method_cov, d=0.94)

## Define the parameters for the portfolio optimization
model='Classic'
rm = 'UCI' # Ulcer Index as risk measure
obj = 'Sharpe' # Maximum Ulcer Performance Index as objective function
hist = True # Use historical scenarios
l = 0 # Risk aversion factor, not used
rf = 0 # Risk free rate

## Run the optimization
w = port_tls.optimization(model=model, rm=rm, obj=obj, rf=rf, l=l, hist=hist)
print(w)

           weights
A     1.902414e-11
AAPL  3.141738e-03
ABC   2.157479e-01
ABMD  2.083451e-02
ABT   1.733748e-01
ADBE  6.354289e-02
ADM   3.219747e-02
ADSK  1.638863e-01
AEE   2.302002e-11
AEP   6.629259e-10
AES   3.365239e-11
AFL   2.388472e-11
AGN   2.442134e-11
AIG   1.440791e-11
AIV   7.403369e-02
AJG   2.500600e-01
AKAM  2.590331e-11
ALB   1.341028e-10
ALK   3.180646e-03
ALL   1.628796e-11
CPU times: user 256 ms, sys: 173 ms, total: 429 ms
Wall time: 344 ms


In [8]:
%%time

# Calculate the optimal portfolio with Portfolio Optimizer
po_request_body = { 'assets': list(map(lambda x: {'assetPrices': x.tolist()}, asset_prices_tls_np)), 
                    'riskFreeRate': rf }
response = requests.post('https://api.portfoliooptimizer.io/v1/portfolio/optimization/maximum-ulcer-performance-index', json=po_request_body)
print(response.json())

{'assetsWeights': [2.818666543917401e-10, 0.0031417761902323964, 0.21574822298607402, 0.020834312504754097, 0.17337499050928162, 0.0635431417322566, 0.03219802833097268, 0.16388631415562008, 6.901618267607307e-10, 7.006597474035169e-09, 1.402220260144518e-10, 2.349219039140339e-10, 2.300208641062024e-10, 3.267086938550072e-10, 0.07403289916107861, 0.2500600811703604, 5.183852911995602e-11, 1.7699000557006747e-09, 0.0031802219036428958, 6.243694404308834e-10]}
CPU times: user 29.5 ms, sys: 3.84 ms, total: 33.3 ms
Wall time: 1.4 s


> Again, the computed portfolio weights pretty much agree.

> Performances wise, Riskfolio-Lib is the clear winner on this particular example, but most of the time taken by Portfolio Optimizer is actually spent in network transfer.

> Let's check this by compressing the input data!

In [18]:
import gzip
import json

# Compress the input data for Portfolio Optimizer
po_request_body = { 'assets': list(map(lambda x: {'assetPrices': x.tolist()}, asset_prices_tls_np)), 
                    'riskFreeRate': rf }
po_request_body_gzip = gzip.compress(json.dumps(po_request_body).encode('utf-8'))

In [19]:
%%time

# Calculate the optimal portfolio with Portfolio Optimizer
response = requests.post('https://api.portfoliooptimizer.io/v1/portfolio/optimization/maximum-ulcer-performance-index', 
                         data = po_request_body_gzip, 
                         headers = {'Content-Encoding': 'gzip', 
                                    'Content-Type' : 'application/json'})
print(response.json())


{'assetsWeights': [2.818666543917401e-10, 0.0031417761902323964, 0.21574822298607402, 0.020834312504754097, 0.17337499050928162, 0.0635431417322566, 0.03219802833097268, 0.16388631415562008, 6.901618267607307e-10, 7.006597474035169e-09, 1.402220260144518e-10, 2.349219039140339e-10, 2.300208641062024e-10, 3.267086938550072e-10, 0.07403289916107861, 0.2500600811703604, 5.183852911995602e-11, 1.7699000557006747e-09, 0.0031802219036428958, 6.243694404308834e-10]}
CPU times: user 26.1 ms, sys: 3.19 ms, total: 29.3 ms
Wall time: 1.01 s


> Thanks to gzip, the total response time of Portfolio Optimizer has been divided by ~33%.

> In addition, and what does not appear here, is that the internal logs of Portfolio Optimizer show a pure computation time of ~500 ms, comparable to the computation time of Riskfolio-lib!